### PR TITLE
[Breaking] Change default evaluation metric for classification to logloss / mlogloss

### LINKED ISSUE #6070: Change default eval_metric for binary:logistic objective + add warning for missing eval_metric when early stopping is enabled
I stumbled over the default metric of the `binary:logistic` objective. It seems to be 1-accuracy, which is a rather unfortunate choice. Accuracy is not even a proper scoring rule, see e.g. [Wiki](https://en.wikipedia.org/wiki/Scoring_rule#Proper_scoring_rules).

In my view, it should be "logloss", which is a strictly proper scoring rule in estimating the expectation under the binary objective.

-- Edit

The problem occurs with early stopping without manually setting the `eval_metric`. The default evaluation metric should at least be a strictly consistent scoring rule. 

I am using R with XGBoost version `1.1.1.1`.

### PR DESCRIPTION
**Pull request by lorentzenchr**
Closes #6070.
Change default evaluation metric for binary classification to `"logloss"`, and for muli-class classification to "`mlogloss`".

### CODE REVIEW

**Comment by hcho3** [comment]
This pull request does not add a warning for performing early stopping with a default metric. So we should keep the issue open.

**Comment by hcho3** [comment]
Let me know if you need help.

**Comment by hcho3** [comment]
@lorentzenchr You should update this line too:
https://github.com/dmlc/xgboost/blob/444131a2e6efd21825b10074e62d2b33735058e6/plugin/example/custom_obj.cc#L58-L60
This should fix the error in the Google C++ test. (We use Google Test framework, so hence the name.)

Also, many unit tests for the R package have assertions that checks the default evaluation metrics, so they need to be updated as well. For example:
https://github.com/dmlc/xgboost/blob/444131a2e6efd21825b10074e62d2b33735058e6/R-package/tests/testthat/test_basic.R#L18-L21
There are many of such occurrences in other R unit tests.

**Comment by hcho3** [comment]
@lorentzenchr 

To make it easier, let's just throw a warning whenever no `eval_metric` is explicitly set, regardless of whether early stopping is enabled. This way, we only have to change a single place in the codebase, as follows:
https://github.com/dmlc/xgboost/blob/dda9e1e4879118738d9f9d5094246692c0f6123c/src/learner.cc#L1033-L1036

If you think this is a good idea, I'll add a commit to this pull request.

**Comment by lorentzenchr** [comment]
@hcho3 Now, it is starting to look good. Thanks a lot for your help!

File: src/learner.cc [hunk 1]
```diff
@@ -1031,6 +1031,18 @@ class LearnerImpl : public LearnerIO {
     std::ostringstream os;
     os << '[' << iter << ']' << std::setiosflags(std::ios::fixed);
     if (metrics_.size() == 0 && tparam_.disable_default_eval_metric <= 0) {
+      auto warn_default_eval_metric = [](const std::string& objective, const std::string& before,
```
Comment (trivialfis) [inline src/learner.cc:1034]: Can we place this warning inside objective function `DefaultEvalMetric`?  Also, are we sure this warning is only emitted once during training?

Comment (hcho3) [reply src/learner.cc:1034]: I don't think moving the warning to `ObjFunction::DefaultEvalMetric()` is a good idea, since then the warning would appear in multiple places. I'd like to keep it here so that we can easily remove it later.

> Also, are we sure this warning is only emitted once during training?

Yes. The warning is emitted just before the default evaluation metric gets added to the vector `metrics_`. Once the default metric is in `metrics_`, the warning will not be thrown.

**Comment by hcho3** [comment]
@mayer79  Can you review?

File: R-package/tests/testthat/test_callbacks.R [hunk 2]
```diff
@@ -236,7 +238,7 @@ test_that("early stopping xgb.train works", {
 test_that("early stopping using a specific metric works", {
   set.seed(11)
   expect_output(
-    bst <- xgb.train(param, dtrain, nrounds = 20, watchlist, eta = 0.6,
+    bst <- xgb.train(param[-2], dtrain, nrounds = 20, watchlist, eta = 0.6,
```
Comment (hcho3) [inline R-package/tests/testthat/test_callbacks.R:241]: What is this for? @lorentzenchr

**Comment by mayer79** [comment]
@hcho3 and @lorentzenchr : I had a look at the changes. They lgtm. I was wondering about the related objective 
"binary:logitraw", which uses "auc" as default. Consistently, this should be "logloss" as well. What do you think?

**Comment by hcho3** [comment]
I prefer to keep the default for binary:logitraw. Is AUC not a proper scoring metric?

File: R-package/tests/testthat/test_callbacks.R (hunk 2 above)
Comment (lorentzenchr) [reply R-package/tests/testthat/test_callbacks.R:241]: param[2] sets eval_metric. I exclude it here, because it is already specified by kwarg. Otherwise it throws a warnung and the test fails. 

Should I add a comment or alternatively remove the kwarg?

**Comment by lorentzenchr** [comment]
> I prefer to keep the default for binary:logitraw. Is AUC not a proper scoring metric?

Unfortunately, not at all. There is no property/functional of the target distribution that is consistently estimated my maximizing AUC (or nobody has found one yet:smirk:).

**Comment by mayer79** [comment]
> I prefer to keep the default for binary:logitraw. Is AUC not a proper scoring metric?

If I am not wrong, binary:logitraw trains the same model as binary:logistic, just without backtransforming the prediction to probability scale?

File: R-package/tests/testthat/test_callbacks.R (hunk 2 above)
Comment (hcho3) [reply R-package/tests/testthat/test_callbacks.R:241]: Got it. Let's keep it there.

**Comment by hcho3** [comment]
Yes, binary:logitraw produces models with logit output. Is logloss a proper scoring rule when output is logit?

**Comment by hcho3** [comment]
FYI, I found this link that calls AUC a "semi-proper" scoring rule:  https://stats.stackexchange.com/questions/339919/what-does-it-mean-that-auc-is-a-semi-proper-scoring-rule

**Comment by lorentzenchr** [comment]
@hcho3 Although I find it a bit inconsistent to have different default values of `eval_metric` for `binary:logitraw` and `binary:logistic` (and `reg:logistic`), most use cases are covered with this PR changing the default only for `binary:logistic`. Let's move forward?

**Comment by hcho3** [comment]
@lorentzenchr Sure, we can always come back to `binary:logitraw` later.
